File: functions/server/src/features/emi/models/emi.model.js
const mongoose = require('mongoose');

const emiSchema = new mongoose.Schema(
  {
    user: { type: mongoose.Schema.Types.ObjectId, ref: 'ClientUser', required: true },

    principalAmount: { type: Number, required: true }, // Base amount before interest
    interestPercentage: { type: Number, required: true }, // Interest/markup percentage
    totalAmount: { type: Number, required: true }, // Principal + (Principal × Interest%)
    
    description: { type: String, required: true }, // EMI description instead of products

    billNumber: { type: String, unique: true, required: true },

    startDate: { type: Date, required: true },
    
    // Payment schedule configuration
    // Required for equal installments, optional for custom schedules
    paymentScheduleType: { 
      type: String, 
      enum: ['1', '3', '6', '9', '12', '18'], // Duration in months
      required: false // Optional for custom schedules
    },
    
    dueDates: [{ 
      type: Number, // Day of month (1-31) - can have multiple dates per month
      min: 1,
      max: 31
    }], // e.g., [5, 15, 25] for 5th, 15th, 25th of each month
    // Note: dueDates can be empty array [] for custom schedules (dates are in paymentSchedule items)
    // Required for equal installments mode

    paidInstallments: { type: Number, default: 0 },
    totalInstallments: { type: Number, required: true }, // Total number of payment installments

    status: {
      type: String,
      enum: ['active', 'completed'],
      default: 'active',
    },

    createdBy: { type: mongoose.Schema.Types.ObjectId, ref: 'Admin', required: true }, // Track which admin created this EMI
  },
  { timestamps: true }
);

module.exports = mongoose.model('EMI', emiSchema);

File: lib/models/emi_model.dart
class EmiModel {
  final String id;
  final String billNumber;
  final double principalAmount;
  final double interestPercentage;
  final double totalAmount;
  final String description;
  final int paidInstallments;
  final int totalInstallments;
  final String status;

  EmiModel({
    required this.id,
    required this.billNumber,
    required this.principalAmount,
    required this.interestPercentage,
    required this.totalAmount,
    required this.description,
    required this.paidInstallments,
    required this.totalInstallments,
    required this.status,
  });

  factory EmiModel.fromJson(Map<String, dynamic> json) {
    return EmiModel(
      id: json["_id"],
      billNumber: json["billNumber"] ?? "",
      principalAmount: (json["principalAmount"] ?? 0).toDouble(),
      interestPercentage: (json["interestPercentage"] ?? 0).toDouble(),
      totalAmount: (json["totalAmount"] ?? 0).toDouble(),
      description: json["description"] ?? "",
      paidInstallments: json["paidInstallments"] ?? 0,
      totalInstallments: json["totalInstallments"] ?? 0,
      status: json["status"] ?? "active",
    );
  }
}
